azurelinuxagent/ga/monitor.py

```diff
@@ -329,10 +330,26 @@ def add_sysinfo(self, event):
         # Add sys_info params populated by the agent
         final_parameters.extend(self.sysinfo)
 
-        # Container id is populated for agent events on the fly, but not for extension events. Add it if it's missing.
-        if "ContainerId" not in [param.name for param in event.parameters]:
+        if "ContainerId" not in event:
+            # ContainerId is populated for agent events on the fly, but not for extension events. Add it if it's missing.
             final_parameters.append(
-                TelemetryEventParam("ContainerId", os.environ.get(CONTAINER_ID_ENV_VARIABLE, "UNINITIALIZED")))
+                TelemetryEventParam("ContainerId", get_container_id_from_env()))
+        if "GAVersion" not in event:
```

Comment:
we should merge this list of parameters with the list in _add_default_parameters_to_event, since they must be the same.

you could, for example, add a parameter to _add_default_parameters_to_event to indicate whethet to set the values uncoditionally (as it currently does) or populate them with default values when they are not present (as this code does)

the idea is to enforce consistency and reduce the chance of somebody forgetting to add a parameter in one of those 2 places

Reply:
Or we can have a static list of parameter names (something like `AGENT_TELEMETRY_PARAMETERS = ["GAVersion", "OpcodeName", ...]` and iterate through it in both places, ensuring all of them are added.

Reply:
List by itself would not help much, I think, since the value is different for each item... or how could we use it? maybe I don't get your idea.

I think what we would like to enforce is that the same  values are set on both places, and that the correct policy (set the current value unconditionally vs set default values) is done in the correct place.

Reply:
I just realized what you were suggesting. I agree that the list wouldn't help since the values are different.
If we add a flag like you're suggesting, then this function can also call `_add_default_parameters_to_event` instead of setting them on its own.

Reply:
yes, sorry I did not explain fully. You are right, the idea is that the list of values and the logic to set them would be in a single place (_add_default_parameters_to_event) instead of 2.

Reply:
I will make _add_default_params into a static function in the Event class with a "defaults" flag. to make it easier.

Reply:
Making it static wasn't easy as the function deals with `event` rather than an `array`. I found a better way of iterating tuples and using it now. Updates.

Reply:
if you prefer to postpone this post-release that would work too

but we should refactor this code